From 11267fe76f81dce283d565d517b679aa2be44466 Mon Sep 17 00:00:00 2001
From: Darren Etheridge <detheridge@ti.com>
Date: Fri, 26 Jan 2024 10:54:49 -0600
Subject: [PATCH] chromium: gpu: sandbox: allow access to PowerVR GPU from
 sandbox

Chromium runs in a sandbox to limit access to the system, however
the PowerVR drivers for the Imagination GPU used on TI hardware need
some extra libraries along with the DRM device nodes to be opened up.
This patch opens up the necessary pieces.

Signed-off-by: Darren Etheridge <detheridge@ti.com>
---
 content/gpu/gpu_sandbox_hook_linux.cc | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/content/gpu/gpu_sandbox_hook_linux.cc b/content/gpu/gpu_sandbox_hook_linux.cc
index d93285a..1f8aafd 100644
--- a/content/gpu/gpu_sandbox_hook_linux.cc
+++ b/content/gpu/gpu_sandbox_hook_linux.cc
@@ -67,6 +67,11 @@ inline bool UseChromecastSandboxAllowlist() {
 #endif
 }

+inline bool IsGPUIMGRogue() {
+  return true;
+}
+
+
 inline bool IsArchitectureArm() {
 #if defined(ARCH_CPU_ARM_FAMILY)
   return true;
@@ -441,6 +446,11 @@ std::vector<BrokerFilePermission> FilePermissionsForGpu(

   AddVulkanICDPermissions(&permissions);

+  if (IsGPUIMGRogue()) {
+      // Add standard DRM permissions for snapdragon/PowerVR:
+      AddDrmGpuPermissions(&permissions);
+  }
+
   if (IsChromeOS()) {
     // Permissions are additive, there can be multiple GPUs in the system.
     AddStandardChromeOsPermissions(&permissions);
@@ -508,6 +518,8 @@ void LoadArmGpuLibraries() {
         DRI_DRIVER_DIR "/mediatek_dri.so",
         DRI_DRIVER_DIR "/rockchip_dri.so",
         DRI_DRIVER_DIR "/asahi_dri.so",
+        DRI_DRIVER_DIR "/pvr_dri.so",
+        DRI_DRIVER_DIR "/tidss_dri.so",
 #else
         "/usr/lib64/dri/msm_dri.so",
         "/usr/lib64/dri/panfrost_dri.so",
@@ -515,6 +527,8 @@ void LoadArmGpuLibraries() {
         "/usr/lib64/dri/rockchip_dri.so",
         "/usr/lib64/dri/asahi_dri.so",
         "/usr/lib/dri/msm_dri.so",
+        "/usr/lib/dri/tidss_dri.so",
+        "/usr/lib/dri/pvr_dri.so",
         "/usr/lib/dri/panfrost_dri.so",
         "/usr/lib/dri/mediatek_dri.so",
         "/usr/lib/dri/rockchip_dri.so",
@@ -632,7 +646,7 @@ sandbox::syscall_broker::BrokerCommandSet CommandSetForGPU(
   command_set.set(sandbox::syscall_broker::COMMAND_ACCESS);
   command_set.set(sandbox::syscall_broker::COMMAND_OPEN);
   command_set.set(sandbox::syscall_broker::COMMAND_STAT);
-  if (IsChromeOS() &&
+  if ((IsGPUIMGRogue() || IsChromeOS()) &&
       (options.use_amd_specific_policies ||
        options.use_intel_specific_policies ||
        options.use_virtio_specific_policies || IsArchitectureArm())) {
--
2.36.1
